astropy/nddata/decorators.py

```diff
@@ -3,161 +3,236 @@
+        if returns is None:
+            returns = []
+        if keeps is None:
+            keeps = []
+
+    # Short version to avoid the long variable name later.
+    attr_arg_map = attribute_argument_mapping
+    if any(keep in returns for keep in keeps):
+        raise ValueError("cannot specify the same attribute in `returns` and "
+                         "`keeps`.")
+    all_returns = returns + keeps
 
     def support_nddata_decorator(func):
-
         # Find out args and kwargs
-        sig = signature(func)
-        func_args = []
-        func_kwargs = []
-        for param in sig.parameters.values():
+        func_args, func_kwargs = [], []
+        sig = signature(func).parameters
+        for param_name, param in six.iteritems(sig):
             if param.kind in (param.VAR_POSITIONAL, param.VAR_KEYWORD):
-                raise ValueError("func may not have *args or **kwargs")
+                raise ValueError("func may not have *args or **kwargs.")
             elif param.default == param.empty:
-                func_args.append(param.name)
+                func_args.append(param_name)
             else:
-                func_kwargs.append(param.name)
+                func_kwargs.append(param_name)
 
         # First argument should be data
-        if len(func_args) == 0 or func_args[0] != 'data':
-            raise ValueError("Can only wrap functions whose first positional argument is `data`")
-
-        supported_properties = ['uncertainty', 'mask', 'meta', 'unit', 'wcs']
+        if not func_args or func_args[0] != attr_arg_map.get('data', 'data'):
+            raise ValueError("Can only wrap functions whose first positional "
+                             "argument is `{0}`"
+                             "".format(attr_arg_map.get('data', 'data')))
 
         @wraps(func)
         def wrapper(data, *args, **kwargs):
-
             unpack = isinstance(data, accepts)
             input_data = data
-
+            ignored = []
             if not unpack and isinstance(data, NDData):
                 raise TypeError("Only NDData sub-classes that inherit from {0}"
-                                " can be used by this function".format(accepts.__name__))
+                                " can be used by this function"
+                                "".format(accepts.__name__))
 
-            # If data is an NDData instance, we can try and find properties that
-            # can be passed as kwargs.
+            # If data is an NDData instance, we can try and find properties
+            # that can be passed as kwargs.
             if unpack:
-
-                ignored = []
-
                 # We loop over a list of pre-defined properties
-                for prop in supported_properties:
-
-                    # We only need to do something if the property exists on the
-                    # NDData object
-                    if hasattr(data, prop):
+                for prop in islice(SUPPORTED_PROPERTIES, 1, None):
```

Comment:
Let's not worry about it given the large number of other things to be done before feature freeze.